File: backend/app/util/coords.py
"""Coordinate utilities.

The GCJ-02 invariant: everything in this project is already GCJ-02 (Amap native), so
NO conversion happens between map, DB and API. The single exception would be
navigator.geolocation (WGS-84); when that lands in a later milestone,
``wgs84_to_gcj02`` is the ONLY place a transform may go, guarded by ``is_in_china()``
(the offset is undefined outside the Chinese grid).
"""

from __future__ import annotations

import math

Coord = tuple[float, float]

_EARTH_RADIUS_M = 6_371_000.0


def haversine_m(a: Coord, b: Coord) -> float:
    """Great-circle distance in metres. The zero-API cost model for TSP scoring."""
    lat1, lat2 = math.radians(a[1]), math.radians(b[1])
    dlat = lat2 - lat1
    dlng = math.radians(b[0] - a[0])
    h = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlng / 2) ** 2
    return 2 * _EARTH_RADIUS_M * math.asin(math.sqrt(h))


# Rough urban speeds for turning a straight-line cost into an ETA. Estimates by
# design -- real road times come from the Amap matrix; this only serves the default
# haversine cost model, where a plausible ordering beats no ordering.
_URBAN_KMH = {"driving": 30.0, "walking": 4.5, "straight": 30.0}


def estimate_seconds(metres: float, mode: str = "driving") -> int:
    kmh = _URBAN_KMH.get(mode, 30.0)
    return int(round(metres / 1000.0 / kmh * 3600.0))


def is_in_china(lng: float, lat: float) -> bool:
    """Coarse Chinese bounding box. Good enough as a transform guard."""
    return 73.0 <= lng <= 136.0 and 17.0 <= lat <= 54.0


def wgs84_to_gcj02(lng: float, lat: float) -> Coord:
    """WGS-84 -> GCJ-02. UNUSED until geolocation support lands; deliberately the only
    transform in the project. Outside China the offset is undefined -> identity."""
    if not is_in_china(lng, lat):
        return lng, lat

    a = 6378245.0  # semi-major axis of the克拉索夫斯基 ellipsoid
    ee = 0.00669342162296594323

    def _t_lat(x: float, y: float) -> float:
        ret = (
            -100.0
            + 2.0 * x
            + 3.0 * y
            + 0.2 * y * y
            + 0.1 * x * y
            + 0.2 * math.sqrt(abs(x))
        )
        ret += (
            20.0 * math.sin(6.0 * x * math.pi) + 20.0 * math.sin(2.0 * x * math.pi)
        ) * 2.0 / 3.0
        ret += (
            20.0 * math.sin(y * math.pi) + 40.0 * math.sin(y / 3.0 * math.pi)
        ) * 2.0 / 3.0
        ret += (
            160.0 * math.sin(y / 12.0 * math.pi) + 320 * math.sin(y * math.pi / 30.0)
        ) * 2.0 / 3.0
        return ret

    def _t_lng(x: float, y: float) -> float:
        ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * math.sqrt(abs(x))
        ret += (
            20.0 * math.sin(6.0 * x * math.pi) + 20.0 * math.sin(2.0 * x * math.pi)
        ) * 2.0 / 3.0
        ret += (
            20.0 * math.sin(x * math.pi) + 40.0 * math.sin(x / 3.0 * math.pi)
        ) * 2.0 / 3.0
        ret += (
            150.0 * math.sin(x / 12.0 * math.pi) + 300.0 * math.sin(x / 30.0 * math.pi)
        ) * 2.0 / 3.0
        return ret

    dlat = _t_lat(lng - 105.0, lat - 35.0)
    dlng = _t_lng(lng - 105.0, lat - 35.0)
    radlat = lat / 180.0 * math.pi
    magic = math.sin(radlat)
    magic = 1 - ee * magic * magic
    sqrtmagic = math.sqrt(magic)
    dlat = (dlat * 180.0) / ((a * (1 - ee)) / (magic * sqrtmagic) * math.pi)
    dlng = (dlng * 180.0) / (a / sqrtmagic * math.cos(radlat) * math.pi)
    return lng + dlng, lat + dlat

File: frontend/src/utils/coords.ts
/**
 * Frontend port of backend/app/util/coords.py.
 *
 * Everything in this project is GCJ-02 (Amap native): map, DB, API -- no conversion
 * anywhere. The ONE exception is navigator.geolocation, which returns WGS-84 and is
 * off by 100-700 m in mainland China. Convert at exactly that boundary and nowhere
 * else. Outside China the offset is undefined -> identity.
 */

export type Coord = [number, number]

const PI = Math.PI
const A = 6378245.0 // 克拉索夫斯基椭球长半轴
const EE = 0.00669342162296594323

export function isInChina(lng: number, lat: number): boolean {
  return lng >= 73.0 && lng <= 136.0 && lat >= 17.0 && lat <= 54.0
}

function transformLat(x: number, y: number): number {
  let ret =
    -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * Math.sqrt(Math.abs(x))
  ret += ((20.0 * Math.sin(6.0 * x * PI) + 20.0 * Math.sin(2.0 * x * PI)) * 2.0) / 3.0
  ret += ((20.0 * Math.sin(y * PI) + 40.0 * Math.sin((y / 3.0) * PI)) * 2.0) / 3.0
  ret += ((160.0 * Math.sin((y / 12.0) * PI) + 320.0 * Math.sin((y * PI) / 30.0)) * 2.0) / 3.0
  return ret
}

function transformLng(x: number, y: number): number {
  let ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * Math.sqrt(Math.abs(x))
  ret += ((20.0 * Math.sin(6.0 * x * PI) + 20.0 * Math.sin(2.0 * x * PI)) * 2.0) / 3.0
  ret += ((20.0 * Math.sin(x * PI) + 40.0 * Math.sin((x / 3.0) * PI)) * 2.0) / 3.0
  ret += ((150.0 * Math.sin((x / 12.0) * PI) + 300.0 * Math.sin((x / 30.0) * PI)) * 2.0) / 3.0
  return ret
}

/** WGS-84 -> GCJ-02. Identity outside the Chinese grid. */
export function wgs84ToGcj02(lng: number, lat: number): Coord {
  if (!isInChina(lng, lat)) return [lng, lat]
  let dlat = transformLat(lng - 105.0, lat - 35.0)
  let dlng = transformLng(lng - 105.0, lat - 35.0)
  const radlat = (lat / 180.0) * PI
  let magic = Math.sin(radlat)
  magic = 1 - EE * magic * magic
  const sqrtmagic = Math.sqrt(magic)
  dlat = (dlat * 180.0) / ((A * (1 - EE)) / (magic * sqrtmagic) * PI)
  dlng = (dlng * 180.0) / ((A / sqrtmagic) * Math.cos(radlat) * PI)
  return [lng + dlng, lat + dlat]
}

const EARTH_R_M = 6371000

/** 大圆距离（米）。GCJ-02 上算，小尺度下与真实路网距离的偏差可忽略——只用于展示。 */
export function haversineM(
  a: { lng: number; lat: number },
  b: { lng: number; lat: number },
): number {
  const rad = PI / 180
  const dLat = (b.lat - a.lat) * rad
  const dLng = (b.lng - a.lng) * rad
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(a.lat * rad) * Math.cos(b.lat * rad) * Math.sin(dLng / 2) ** 2
  return 2 * EARTH_R_M * Math.asin(Math.sqrt(s))
}

/** 展示用距离文本：<1km 显示米，其余 1 位小数公里。 */
export function formatDistance(m: number): string {
  if (m < 1000) return `${Math.round(m / 10) * 10} 米`
  return `${m / 1000 >= 10 ? Math.round(m / 1000) : (m / 1000).toFixed(1)} 公里`
}
